linkerd/drain/src/retain.rs

```diff
@@ -0,0 +1,70 @@
+use crate::Watch;
+use linkerd2_stack::{layer, NewService};
+use std::{
+    future::Future,
+    pin::Pin,
+    task::{Context, Poll},
+};
+
+/// Holds a drain::Watch for as long as a request is pending.
+#[derive(Clone, Debug)]
+pub struct NewRetain<N> {
+    inner: N,
+    drain: Watch,
```

Comment:
since this *also* holds the `Watch`, it will keep the drain from completing as long as a stack with a `NewRetain` in it exists. This may not be an issue if signalling a drain also cancels the tasks which hold these stacks (which I think it will), but this might be worth being aware of for the future...